stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -53,59 +54,115 @@ use crate::util_lib::db::Error as DBError;
 pub mod tests;
 
 impl<'a, T: BlockEventDispatcher> OnChainRewardSetProvider<'a, T> {
-    pub fn get_reward_set_nakamoto(
+    /// Read a reward_set written while updating .signers
+    /// `debug_log` should be set to true if the reward set loading should
+    ///  log messages as `debug!` instead of `error!` or `info!`. This allows
+    ///  RPC endpoints to expose this without flooding loggers.
+    pub fn read_reward_set_nakamoto(
         &self,
         cycle_start_burn_height: u64,
         chainstate: &mut StacksChainState,
         burnchain: &Burnchain,
         sortdb: &SortitionDB,
         block_id: &StacksBlockId,
+        debug_log: bool,
     ) -> Result<RewardSet, Error> {
-        // TODO: this method should read the .signers contract to get the reward set entries.
-        //   they will have been set via `NakamotoChainState::check_and_handle_prepare_phase_start()`.
         let cycle = burnchain
             .block_height_to_reward_cycle(cycle_start_burn_height)
             .expect("FATAL: no reward cycle for burn height");
+        // figure out the block ID
+        let Some(coinbase_height_of_calculation) = chainstate
+            .eval_boot_code_read_only(
```

Comment:
Doesn't have to be done now, but this is going to be relatively slow while being on a network hot-path.  Is there a way this could be cached?

Reply:
Yeah -- if we want to cache this in the network path, we could certainly do so there (or cache the whole struct).

Reply:
Yeah, was thinking we'd cache the whole struct.  Can do this in a separate PR though.